---
title: "Class19"
author: "Lucy Wang"
format: pdf
---

# Mini-Project: Investigating Pertussis Resurgence

## 1. Investigating pertussis cases by year

```{r}
# install.packages("datapasta")
# Tools -> Addin -> “Paste as data.frame”
cdc <- data.frame(
                                                                    Year = c(1922L,1923L,1924L,
                                                                             1925L,1926L,1927L,
                                                                             1928L,1929L,
                                                                             1930L,1931L,1932L,
                                                                             1933L,1934L,1935L,
                                                                             1936L,1937L,1938L,
                                                                             1939L,1940L,1941L,
                                                                             1942L,1943L,1944L,
                                                                             1945L,1946L,1947L,
                                                                             1948L,1949L,
                                                                             1950L,1951L,1952L,
                                                                             1953L,1954L,1955L,
                                                                             1956L,1957L,1958L,
                                                                             1959L,1960L,1961L,
                                                                             1962L,1963L,1964L,
                                                                             1965L,1966L,1967L,
                                                                             1968L,1969L,
                                                                             1970L,1971L,1972L,
                                                                             1973L,1974L,1975L,
                                                                             1976L,1977L,1978L,
                                                                             1979L,1980L,1981L,
                                                                             1982L,1983L,1984L,
                                                                             1985L,1986L,1987L,
                                                                             1988L,1989L,
                                                                             1990L,1991L,1992L,
                                                                             1993L,1994L,1995L,
                                                                             1996L,1997L,1998L,
                                                                             1999L,2000L,2001L,
                                                                             2002L,2003L,2004L,
                                                                             2005L,2006L,2007L,
                                                                             2008L,2009L,
                                                                             2010L,2011L,2012L,
                                                                             2013L,2014L,2015L,
                                                                             2016L,2017L,2018L,
                                                                             2019L),
                                            No..Reported.Pertussis.Cases = c(107473,164191,
                                                                             165418,152003,202210,
                                                                             181411,161799,
                                                                             197371,166914,172559,
                                                                             215343,179135,
                                                                             265269,180518,147237,
                                                                             214652,227319,
                                                                             103188,183866,222202,
                                                                             191383,191890,
                                                                             109873,133792,109860,
                                                                             156517,74715,69479,
                                                                             120718,68687,
                                                                             45030,37129,60886,
                                                                             62786,31732,28295,
                                                                             32148,40005,14809,
                                                                             11468,17749,17135,
                                                                             13005,6799,7717,
                                                                             9718,4810,3285,
                                                                             4249,3036,3287,1759,
                                                                             2402,1738,1010,
                                                                             2177,2063,1623,
                                                                             1730,1248,1895,2463,
                                                                             2276,3589,4195,
                                                                             2823,3450,4157,
                                                                             4570,2719,4083,6586,
                                                                             4617,5137,7796,
                                                                             6564,7405,7298,
                                                                             7867,7580,9771,
                                                                             11647,25827,25616,
                                                                             15632,10454,13278,
                                                                             16858,27550,18719,
                                                                             48277,28639,32971,
                                                                             20762,17972,18975,
                                                                             15609,18617)
                                          )
```

> Q1. With the help of the R “addin” package datapasta assign the CDC pertussis case number data to a data frame called cdc and use ggplot to make a plot of cases numbers over time.

```{r}
library(ggplot2)
p <- ggplot(cdc) +
  aes(Year, No..Reported.Pertussis.Cases) +
  geom_point() +
  geom_line() +
  labs(title = "Pertussis Cases by Year (1922-2019)", x = "Year", y = "Number of Cases")

p
```

## 2. A tale of two vaccines (wP & aP)

Examine what happened after the switch to the acellular pertussis (aP) vaccination program, adding lines to plot.

> Q2. Using the ggplot geom_vline() function add lines to your previous plot for the 1946 introduction of the wP vaccine and the 1996 switch to aP vaccine (see example in the hint below). What do you notice?

```{r}
# Adding geom_vline (xintercept = 1946 & 1996)
p + 
  geom_vline(xintercept = 1946, color = "red", linetype = 2) +
  geom_vline(xintercept = 1996, color = "blue", linetype = 2)
  

```

> Q3. Describe what happened after the introduction of the aP vaccine? Do you have a possible explanation for the observed trend?

There is a raise in number of cases after the introduction of aP vaccine. It is possible that due to more frequent testing, the disease is diagnosed more frequently. Maybe the pathogen evolved to escape some of human immune mechanisms from decades of attempts to invade human.

## 3. Exploring CMI-PB data

```{r}
# Allows us to read, write and process JSON data
library(jsonlite)
```

```{r}
# Read table from CMI-PB API
subject <- read_json("https://www.cmi-pb.org/api/subject", simplifyVector = TRUE) 
```

> Q4. How may aP and wP infancy vaccinated subjects are in the dataset?

```{r}
table(subject$infancy_vac)
```

> Q5. How many Male and Female subjects/patients are in the dataset?

```{r}
table(subject$biological_sex)
```


> Q6. What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc…)?

```{r}
table(subject$biological_sex, subject$race)
```

### Side-Note: Working with dates

```{r}
# There is a package written for dates operations
library(lubridate)
today()
```

```{r}
# To Calculate time differences
today() - ymd("2000-01-01")  # In days
time_length( today() - ymd("2000-01-01"),  "years") # In years
```

> Q7. Using this approach determine (i) the average age of wP individuals, (ii) the average age of aP individuals; and (iii) are they significantly different?

```{r}
# Use todays date to calculate age in days
subject$age <- today() - ymd(subject$year_of_birth)
time_length(subject$age, "years")
```

```{r}
# aP average ages
library(dplyr)

ap <- subject %>% filter(infancy_vac == "aP")

round( summary( time_length( ap$age, "years" ) ) )
```

```{r}
# wP average ages
wp <- subject %>% filter(infancy_vac == "wP")
round( summary( time_length( wp$age, "years" ) ) )
```

They are significantly different.

> Q8. Determine the age of all individuals at time of boost?

```{r}
int <- ymd(subject$date_of_boost) - ymd(subject$year_of_birth)
age_at_boost <- time_length(int, "year")
head(age_at_boost)
```

> Q9. With the help of a faceted boxplot (see below), do you think these two groups are significantly different?

```{r}
ggplot(subject) +
  aes(time_length(age, "year"),
      fill=as.factor(infancy_vac)) +
  geom_histogram(show.legend=FALSE) +
  facet_wrap(vars(infancy_vac), nrow=2) 
```

Yes, I think it is significantly different since there is little overlap between two set of data, which gives very different mean and median.

### Joining multiple tables

```{r}
specimen <- read_json("https://www.cmi-pb.org/api/specimen", simplifyVector = TRUE) 
titer <- read_json("https://www.cmi-pb.org/api/ab_titer", simplifyVector = TRUE) 
```

> Q9. Complete the code to join specimen and subject tables to make a new merged data frame containing all specimen records along with their associated subject details:

```{r}
meta <- full_join(specimen, subject)
dim(meta)
head(meta)
```

> Q10. Now using the same procedure join meta with titer data so we can further analyze this data in terms of time of visit aP/wP, male/female etc.

```{r}
abdata <- inner_join(titer, meta, by = "specimen_id")
dim(abdata)
```

> Q11. How many specimens (i.e. entries in abdata) do we have for each isotype?

```{r}
# Use table() on isotype column
table(abdata$isotype)
```

> Q12. What do you notice about the number of visit 8 specimens compared to other visits?

```{r}
table(abdata$visit)
```

8 specimens are far less than the other visits. 


## 4. Examine IgG1 Ab titer levels

```{r}
ig1 <- abdata %>% filter(isotype == "IgG1", visit!=8)
head(ig1)
```

> Q13. Complete the following code to make a summary boxplot of Ab titer levels for all antigens:

```{r}
ggplot(ig1) +
  aes(MFI_normalised, antigen) +
  geom_boxplot() + 
  facet_wrap(vars(visit), nrow=2)
```

> Q14. What antigens show differences in the level of IgG1 antibody titers recognizing them over time? Why these and not others?

Mainly FIM2/3, also FHA and DT show slight increase. Maybe these are epitopes of pathogenic characteristic, so IgG1 could recognize them overtime and become increasingly accurate. 

We can attempt to examine differences between wP and aP here by setting color and/or facet values of the plot to include `infancy_vac` status. However these plots tend to be rather busy and thus hard to interpret easily.

> Q15. Filter to pull out only two specific antigens for analysis and create a boxplot for each. You can chose any you like. Below I picked a “control” antigen (“Measles”, that is not in our vaccines) and a clear antigen of interest (“FIM2/3”, extra-cellular fimbriae proteins from B. pertussis that participate in substrate attachment).

```{r}
# For measles
filter(ig1, antigen=="Measles") %>%
  ggplot() +
  aes(MFI, col=infancy_vac) +
  geom_boxplot(show.legend = "Measles") +
  facet_wrap(vars(visit)) +
  theme_bw()
```

```{r}
# For FIM2/3
filter(ig1, antigen=="FIM2/3") %>%
  ggplot() +
  aes(MFI, col=infancy_vac) +
  geom_boxplot(show.legend = "FIM2/3") +
  facet_wrap(vars(visit)) +
  theme_bw()

```

> Q16. What do you notice about these two antigens time course and the FIM2/3 data in particular?

FIM2/3 rapidly increases from visit 1 and is far more than measles. It reaches its maximum value in visit 5.

> Q17. Do you see any clear difference in aP vs. wP responses?

wP seems to show less a trend of increase and stays lower when aP increases. Maybe indicating that aP is more responsive during the visits. 


## 5. Obtaining CMI-PB RNASeq data

#### Investigating IgG1 gene

```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"

rna <- read_json(url, simplifyVector = TRUE) 
```

```{r}
#Just like meta <- inner_join(specimen, subject)
ssrna <- inner_join(rna, meta)
```

> Q18. Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of `visit` vs. `tpm`).

```{r}
ggplot(ssrna) +
  aes(visit, tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
```

> Q19.: What do you notice about the expression of this gene (i.e. when is it at it’s maximum level)?

It peaks at visit 4. 

> Q20. Does this pattern in time match the trend of antibody titer data? If not, why not?

I think it matches the antibody titer, though the antibody peaks at visit 5 instead of visit 4. We can regard the production of antibody a response to the expression of those genes, which can take 1 visit.


# -The End - 



















